In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/best_rf_1/scikitlearn/default/11/model_rf_3_1.joblib
/kaggle/input/best_rf_1/scikitlearn/default/11/model_rf_3.joblib
/kaggle/input/best_rf_1/scikitlearn/default/11/model_rf_2.joblib
/kaggle/input/best_rf_1/scikitlearn/default/11/model_rf_1_gridsearch.joblib
/kaggle/input/best_rf_1/scikitlearn/default/11/model_rf_4.joblib
/kaggle/input/best_rf_1/scikitlearn/default/12/model_rf_3_1.joblib
/kaggle/input/best_rf_1/scikitlearn/default/12/model_rf_3.joblib
/kaggle/input/best_rf_1/scikitlearn/default/12/model_rf_2.joblib
/kaggle/input/best_rf_1/scikitlearn/default/12/model_svc_linear_1.pkl
/kaggle/input/best_rf_1/scikitlearn/default/12/model_rf_1_gridsearch.joblib
/kaggle/input/best_rf_1/scikitlearn/default/12/model_rf_4.joblib
/kaggle/input/task001-p03-ml-models/scikitlearn/default/1/model_rf_1.joblib
/kaggle/input/task001-p03-ml-models/scikitlearn/default/1/model_rf_10.joblib
/kaggle/input/task001-p03-ml-models/scikitlearn/default/1/model_rf_3.joblib
/kaggle/input/task001-p03

<div>
    <h1> <a id="Obj1">🎯Objectives</a></h1>
    <h3 style="text-align:justify; font-family: Calibri">Build a predictive model that answers the question: “what sorts of people were more likely to survive?” using passenger data (ie name, age, gender, socio-economic class, etc). <br></h3>  
  
<p style = "font-family: Arial; font-size: 14px">Note: This project's dataset was created for pedagogical purposes and may not be indicative of the task .</p>
<hr>

<div class="alert alert-block alert-warning">
<img style="float: left" src="https://i.postimg.cc/kXz8cFqC/005-Img-Yellow-Notes-Draft-1-20220819.png" width="60">
<b style = "font-family: Arial; font-size: 16px">Note:</b><p style = "font-family:Verdana; font-size:14px">We must maintain objectivity when analyzing the data to gain valuable insights. This involves collecting, fact-checking, and challenging the data and other sources. Adopting a Genchi Genbutsu approach, which involves go-seeing for yourself or discussing with subject matter experts, allows us to delve into the data to gain firsthand understanding.</p>
    <p>Data should adhere to the <b>COV-FAST</b> principles: Clear, Objective, Valuable, Focused, Agile, Scientific, and Time-bound and Trustworthiness. Additionally, consider exploring alternative models such as logistic regression, random forest, neural networks, or others.</p>
</div>
    <h3>Framework and methodologies</h3>
     A <b>methodology</b> offers <b>specific steps</b> for completing a project, while a <b>framework</b> provides broader guidance on tools, resources, and workflow processes. As an experienced Green Belt Lean Six Sigma professional, the methodology that I often use is DMAIC. I use PACE as a framework as I found it similar to <b>Knowledge Discovery in Databases (KDD)</b>. Both <b>PACE</b> and <b>KDD</b> aim to guide structured data projects. <b>PACE</b> specifically focuses on machine learning projects, whereas <b>KDD</b> covers a broader range of data discovery techniques.
     
<h3>Data analysis PACE steps:</h3>
   <ol style="font-family:Verdana; font-size:16px">
    <li><img style="float:left" src="https://i.imgur.com/gIne5bH.png" width="50"> Plan/Prepare - import the relevant libraries and data</li> 
    <blockquote>
    <ol>Align project with business needs, requirements and contraints. Select an approriate machine learning model based on the problem and business context. KDD: Selection, and Data Wrangling (Pre-processing and Transformation).
    </ol>
    </blockquote>
        <li><img style="float:left" src="https://i.imgur.com/rb8V6X5.png" width="50">Analyze - EDA</li>
    <blockquote>
    <ol> Understanding data for accurate predictions, focus on the response variable (what the model predicts) and leverage exploratory data analysis to uncover patterns and address irregularities. KDD: Data Mining.
     </ol>
    </blockquote>
    <li><img style="float:left" src="https://i.imgur.com/J4M3HKM.png" width="50">Construct - model </li>
    <blockquote>
    <ol>Construct and evaluate model. KDD: Evaluation.
     </ol>
    </blockquote>
    <li><img style="float:left" src="https://i.imgur.com/wpcEXQC.png" width="50">Execute - share</li>
    <blockquote>
    <ol>Interpret model and share the story. KDD: Communicate to stakeholders.
     </ol>
    </blockquote>

<a id="top"></a>
<h2>Table of contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">

<ol>
    <li><a href="#Obj1">Objectives</a></li>
    <li><a href="#Plan">Plan - import the relevant libraries and data</a></li>
    <ul>
        <li><a href="#Def">Data Info</a></li>
    </ul>
    <li><a href="#EDA">Explarotary Data Analysis (EDA)</a></li>
    <ul>
        <li><a href="#EDA_1">Explarotary Data Analysis (EDA)</a></li>
    </ul>
    <li><a href="#Construct">Construct - model</a></li>
    <ul>
        <li><a href="#RF">Random forest</a></li>
        <li><a href="#SVC">SVC (Support Vector Classifier)</a></li>
    </ul>
    <li><a href="#Exe">Execute - share</a></li>
    <ul>
        <li><a href="#Sum">Summary </a></li>
    </ul>
</ol>
</div>

<img style="float:left" src="https://i.imgur.com/gIne5bH.png" width="50">

<h1><a id="Plan">Plan -import the relevant libraries and data</a></h1> 

In [2]:
# For linear algebra and data processing (available in kaggle)
#import numpy as np # linear algebra
#import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Data visualization
import matplotlib.pyplot as plt
from matplotlib.lines import Line2D
import seaborn as sns

# Display full dataframe columns
pd.set_option('display.max_columns', None)

# Regular expressions and statistics
import re
from scipy import stats

# Machine learning models
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier, plot_tree
from sklearn.ensemble import (
    RandomForestClassifier,
    AdaBoostClassifier,
    ExtraTreesClassifier,
    VotingClassifier,
    GradientBoostingClassifier
)
from sklearn.neighbors import KNeighborsClassifier
from xgboost import XGBClassifier, plot_importance

# Preprocessing, model selection, metrics
from sklearn.model_selection import (
    train_test_split, GridSearchCV, cross_val_score,
    StratifiedKFold, cross_validate, cross_val_predict
)
from sklearn.preprocessing import StandardScaler, OneHotEncoder, OrdinalEncoder
from sklearn.impute import SimpleImputer
from sklearn.metrics import (
    accuracy_score, precision_score, recall_score, f1_score,
    confusion_matrix, ConfusionMatrixDisplay, classification_report,
    roc_auc_score, roc_curve
)
from sklearn.utils.class_weight import compute_sample_weight
from sklearn.feature_selection import SelectFromModel
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer

# Imbalanced learning
#from imblearn.pipeline import Pipeline as ImbPipeline
#from imbalanced-learn.pipeline import Pipeline
#from imblearn.over_sampling import SMOTE

# SHAP for model explainability
import shap

# Saving/loading models
import joblib
from joblib import dump, load

# Miscellaneous
from collections import Counter

<h2><a id="Def">Data info</a></h2>
<h3 style="text-align:justify; font-family: Calibri"><br> Train.csv will contain the details of a subset of the passengers on board (891 to be exact) and importantly, will reveal whether they survived or not, also known as the “ground truth”. <br> 
    <br>  The test.csv dataset contains similar information but does not disclose the “ground truth” for each passenger. It’s your job to predict these outcomes.  <br> 
   <br>   Using the patterns you find in the train.csv data, predict whether the other 418 passengers on board (found in test.csv) survived.  <br> 
  <br>    Check out the “Data” tab to explore the datasets even further. Once you feel you’ve created a competitive model, submit it to Kaggle to see where your model stands on our leaderboard against other Kagglers.</h3>

In [3]:
data_df = pd.read_csv('/kaggle/input/task001-p01-ds-titanic-20250411/train_df_1.csv')
all_data_df = pd.read_csv('/kaggle/input/task001-p01-ds-titanic-20250411/all_data_df.csv')

In <a href="https://www.kaggle.com/code/wahyuardhitama/task001-p001-ml-titanic-20250407">part 1 (P01)</a>, we conduct the P and A phase. In <a href="https://www.kaggle.com/code/wahyuardhitama/task001-p002-ml-titanic-20250416">part 2 (P02)</a>, we explore 11 machine learning models.

In [4]:
df=all_data_df.copy()
df.sample(5)

invalid value encountered in greater
invalid value encountered in less
invalid value encountered in greater


PassengerId  Survived  Pclass  \
266           267       0.0       3   
638           639       0.0       3   
1011         1012       NaN       2   
909           910       NaN       3   
1196         1197       NaN       1   

                                                   Name     Sex   Age  SibSp  \
266                           Panula, Mr. Ernesti Arvid    male  16.0      4   
638              Panula, Mrs. Juha (Maria Emilia Ojala)  female  41.0      0   
1011                               Watt, Miss. Bertha J  female  12.0      0   
909                        Ilmakangas, Miss. Ida Livija  female  27.0      1   
1196  Crosby, Mrs. Edward Gifford (Catherine Elizabe...  female  64.0      1   

      Parch            Ticket     Fare Embarked  train_set     AgeGroup  \
266       1           3101295  39.6875        S          1         Teen   
638       5           3101295  39.6875        S          1        Adult   
1011      0        C.A. 33595  15.7500        S          0         Teen   
909       0  STON/O2. 3101270   7.9250        S          0  Young Adult   
1196      1            112901  26.5500        S          0       Senior   

     CabinGroup  CabinNo  CabinCnt  Fare_log Title     Surname TicketGroup  \
266           U        0         0  3.705921    Mr      Panula         XXX   
638           U        0         0  3.705921   Mrs      Panula         XXX   
1011          U        0         0  2.818398  Miss        Watt         SOC   
909           U        0         0  2.188856  Miss  Ilmakangas         XXX   
1196          B       26         1  3.316003   Mrs      Crosby         XXX   

       FamilyGroup FareGroup  FamilySize  
266   Large Family       <50           6  
638   Large Family       <50           6  
1011  Small Family    50-100           3  
909   Small Family       <50           2  
1196  Small Family     >=100           4

In [5]:
df.isna().sum()

PassengerId      0
Survived       418
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             1
Embarked         0
train_set        0
AgeGroup         1
CabinGroup       0
CabinNo          0
CabinCnt         0
Fare_log         1
Title            0
Surname          0
TicketGroup      0
FamilyGroup      0
FareGroup        0
FamilySize       0
dtype: int64

#### Quick fix `Age` and `Fare`

In [6]:
df[df['AgeGroup'].isnull()].head()

PassengerId  Survived  Pclass                                  Name  \
630          631       1.0       1  Barkworth, Mr. Algernon Henry Wilson   

      Sex   Age  SibSp  Parch Ticket  Fare Embarked  train_set AgeGroup  \
630  male  80.0      0      0  27042  30.0        S          1      NaN   

    CabinGroup  CabinNo  CabinCnt  Fare_log Title    Surname TicketGroup  \
630          A       23         1  3.433987    Mr  Barkworth         XXX   

    FamilyGroup FareGroup  FamilySize  
630        Solo       <50           1

In [7]:
df['AgeGroup'].unique()

array(['Young Adult', 'Adult', 'Senior', 'Child', 'Teen', nan],
      dtype=object)

In [8]:
# Fill missing values in 'AgeGroup' with 'Senior' for df
df['AgeGroup'] = df['AgeGroup'].fillna('Senior')

# Do the same for all_data_df
all_data_df['AgeGroup'] = all_data_df['AgeGroup'].fillna('Senior')

In [9]:
# Fill missing 'Fare' values with its median
fare_median = df['Fare'].median()
df['Fare'] = df['Fare'].fillna(fare_median)

# Fill missing 'Fare_log' values with its median
fare_log_median = df['Fare_log'].median()
df['Fare_log'] = df['Fare_log'].fillna(fare_log_median)

# For all_data_df
fare_median_all = all_data_df['Fare'].median()
all_data_df['Fare'] = all_data_df['Fare'].fillna(fare_median_all)

fare_log_median_all = all_data_df['Fare_log'].median()
all_data_df['Fare_log'] = all_data_df['Fare_log'].fillna(fare_log_median_all)

In [10]:
df.select_dtypes(include=[np.number]).columns.tolist()

['PassengerId',
 'Survived',
 'Pclass',
 'Age',
 'SibSp',
 'Parch',
 'Fare',
 'train_set',
 'CabinNo',
 'CabinCnt',
 'Fare_log',
 'FamilySize']

[go to top](#top)  

<img style="float:left" src="https://i.imgur.com/rb8V6X5.png" width="50">

<h1><a id="EDA">Analyze - Explarotary Data Analysis (EDA)</a></h1> 

In <a href="https://www.kaggle.com/code/wahyuardhitama/task001-p001-ml-titanic-20250407">part 1 (P01)</a>, we conduct the P and A phase. In <a href="https://www.kaggle.com/code/wahyuardhitama/task001-p002-ml-titanic-20250416">part 2 (P02)</a>, we explore 11 machine learning models.

### Create function

In [11]:
def quick_view_model(X_train, y_train, cv=5, sort_by='Accuracy'):
    """
    Evaluates multiple classification models using cross-validation.
    
    Parameters:
    - X_train: Training features
    - y_train: Training labels
    - cv: Number of cross-validation folds (default=5)
    - sort_by: Metric to sort the result table by (default='Accuracy')
    
    Returns:
    - results_df: DataFrame with mean and std CV scores for each model
    """

    # Define models
    models = {
        'LogisticRegression': LogisticRegression(max_iter=1000),
        'GaussianNB': GaussianNB(),
        'SVC': SVC(probability=True),
        'DecisionTree': DecisionTreeClassifier(random_state=42),
        'RandomForest': RandomForestClassifier(random_state=42),
        'AdaBoost': AdaBoostClassifier(random_state=42),
        'ExtraTrees': ExtraTreesClassifier(random_state=42),
        'GradientBoosting': GradientBoostingClassifier(random_state=42),
        'XGBoost': XGBClassifier(use_label_encoder=False, eval_metric='logloss', random_state=42),
        'KNN': KNeighborsClassifier()
    }

    # Initialize results dictionary
    results = {}

    # Evaluate each model
    for name, model in models.items():
        pipe = Pipeline([
            ('scaler', StandardScaler()),  # Can skip for tree-based
            ('classifier', model)
        ])

        # Cross-validation
        scores = cross_validate(
            pipe, X_train, y_train, cv=cv,
            scoring={
                'accuracy': 'accuracy',
                'auc': 'roc_auc',
                'f1': 'f1',
                'precision': 'precision',
                'recall': 'recall'
            },
            return_train_score=False
        )

        # Store mean and std for each metric
        results[name] = {
            'Accuracy Mean': np.mean(scores['test_accuracy']),
            'Accuracy Std': np.std(scores['test_accuracy']),
            'AUC Mean': np.mean(scores['test_auc']),
            'AUC Std': np.std(scores['test_auc']),
            'F1-Score Mean': np.mean(scores['test_f1']),
            'F1-Score Std': np.std(scores['test_f1']),
            'Precision Mean': np.mean(scores['test_precision']),
            'Precision Std': np.std(scores['test_precision']),
            'Recall Mean': np.mean(scores['test_recall']),
            'Recall Std': np.std(scores['test_recall'])
        }

    # Convert to DataFrame
    results_df = pd.DataFrame(results).T

    # Sort by selected mean metric
    if f'{sort_by} Mean' in results_df.columns:
        results_df = results_df.sort_values(by=f'{sort_by} Mean', ascending=False)

    # Round for display
    print(f"\nModel Comparison (sorted by {sort_by} Mean):\n")
    print(results_df.round(4))
    
    return results_df

In [12]:
def quick_view_models_plot(results_df, metric='Accuracy'):
    """
    Plots the mean and standard deviation of model performance scores.
    
    Parameters:
    - results_df: DataFrame output from quick_view_model()
    - metric: Metric to plot (e.g., 'Accuracy', 'F1-Score')
    """
    # Construct column names for mean and std
    mean_col = f"{metric} Mean"
    std_col = f"{metric} Std"

    # Check if metric exists
    if mean_col not in results_df.columns or std_col not in results_df.columns:
        raise ValueError(f"Metric '{metric}' not found in results_df. Check column names.")

    # Drop GaussianNB if present
    filtered_df = results_df.drop(index='GaussianNB', errors='ignore')

    # Plotting data
    models = filtered_df.index
    mean_scores = filtered_df[mean_col]
    std_devs = filtered_df[std_col]
    x_pos = np.arange(len(models))

    # Plot
    plt.figure(figsize=(12, 8))
    plt.errorbar(x_pos, mean_scores, yerr=std_devs, fmt='o-', linewidth=2, capsize=6,
                 markersize=8, label=f'Mean {metric} ± Std Dev')
    plt.fill_between(x_pos, mean_scores - std_devs, mean_scores + std_devs,
                     alpha=0.2, color='skyblue')

    plt.xticks(x_pos, models, rotation=45, ha='right')
    plt.grid(True, linestyle='--', alpha=0.7)
    plt.xlabel('Models', fontsize=12, fontweight='bold')
    plt.ylabel(f'{metric} Score', fontsize=12, fontweight='bold')
    plt.title(f'Model Comparison (Excl. GaussianNB) - {metric}', fontsize=14, fontweight='bold')

    # Annotate each point
    for i, (m, s) in enumerate(zip(mean_scores, std_devs)):
        plt.text(i, m + 0.002, f"{m:.3f}±{s:.3f}", ha='center', fontsize=9)

    plt.ylim(min(mean_scores - std_devs) - 0.02, max(mean_scores + std_devs) + 0.02)
    plt.legend(loc='lower right')
    plt.tight_layout()
    plt.show()

In [13]:
def quick_view_models_chart(results_df, metrics=None):
    """
    Plots a grouped bar chart to compare multiple evaluation metrics (Mean) across models.
    """
    if metrics is None:
        metrics = ['Accuracy Mean', 'AUC Mean', 'F1-Score Mean', 'Precision Mean', 'Recall Mean']

    # Drop GaussianNB if present
    #results_df = results_df.drop(index='GaussianNB', errors='ignore')
    models = results_df.index
    x = np.arange(len(models))
    width = 0.13

    colors = ['skyblue', 'orange', 'lightgreen', 'salmon', 'violet']
    if len(metrics) > len(colors):
        raise ValueError("Too many metrics. Add more colors.")

    plt.figure(figsize=(16, 8))

    for i, metric in enumerate(metrics):
        if metric not in results_df.columns:
            raise ValueError(f"Metric '{metric}' not found in results_df.")
        values = results_df[metric]
        plt.bar(x + i * width, values, width, label=metric, color=colors[i])

        for j, val in enumerate(values):
            plt.text(j + i * width, val + 0.015, f"{val:.2f}", ha='center', fontsize=8)

    plt.ylabel('Score', fontsize=12, fontweight='bold')
    plt.title('Model Comparison by Evaluation Metrics (Mean)', fontsize=14, fontweight='bold')
    plt.xticks(x + width * (len(metrics)-1) / 2, models, rotation=45, ha='right')
    plt.ylim(0, 1.1)
    plt.grid(axis='y', linestyle='--', alpha=0.5)
    plt.legend(title="Metrics", loc='upper right')
    plt.tight_layout()
    plt.show()

[go to top](#top)  

<img style="float:left" src="https://i.imgur.com/J4M3HKM.png" width="50">

<h1><a id="Construct">Construct - model</a></h1>  

### Baseline feature selection (original 83 Features, lightly imbalanced data)

In [14]:
#X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)

In [15]:
# Convert the 'Pclass' and 'FamilySize' columns to string (object) type for pd.get_dummies() Using this will acquire 83 features
# Let us apply without converting to string
df[['Pclass', 'FamilySize']] = df[['Pclass', 'FamilySize']].astype(str)

In [16]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Embarked', 'train_set', 'AgeGroup',
       'CabinGroup', 'CabinNo', 'CabinCnt', 'Fare_log', 'Title', 'Surname',
       'TicketGroup', 'FamilyGroup', 'FareGroup', 'FamilySize'],
      dtype='object')

In [17]:
# Created dummy variablesfrom categories (also can use OneHotEncoder)
df_dummies = pd.get_dummies(df[['Pclass', 'Sex', 'Age', 'SibSp',
       'Parch', 'Embarked', 'train_set',
       'CabinGroup', 'CabinCnt', 'Fare_log', 'Title',
       'TicketGroup', 'FamilyGroup','FamilySize']])

X_train = df_dummies[df_dummies.train_set == 1].drop(['train_set'],axis=1)
X_test = df_dummies[df_dummies.train_set == 0].drop(['train_set'], axis=1)

y_train = df[df.train_set==1].Survived
y_train.shape

(891,)

In [18]:
X_train.head()

Age  SibSp  Parch  CabinCnt  Fare_log  Pclass_1  Pclass_2  Pclass_3  \
0  22.0      1      0         0  2.110213     False     False      True   
1  38.0      1      0         1  4.280593      True     False     False   
2  26.0      0      0         0  2.188856     False     False      True   
3  35.0      1      0         1  3.990834      True     False     False   
4  35.0      0      0         0  2.202765     False     False      True   

   Sex_female  Sex_male  Embarked_C  Embarked_Q  Embarked_S  CabinGroup_A  \
0       False      True       False       False        True         False   
1        True     False        True       False       False         False   
2        True     False       False       False        True         False   
3        True     False       False       False        True         False   
4       False      True       False       False        True         False   

   CabinGroup_B  CabinGroup_C  CabinGroup_D  CabinGroup_E  CabinGroup_F  \
0         False         False         False         False         False   
1         False          True         False         False         False   
2         False         False         False         False         False   
3         False          True         False         False         False   
4         False         False         False         False         False   

   CabinGroup_G  CabinGroup_T  CabinGroup_U  Title_Capt  Title_Col  Title_Don  \
0         False         False          True       False      False      False   
1         False         False         False       False      False      False   
2         False         False          True       False      False      False   
3         False         False         False       False      False      False   
4         False         False          True       False      False      False   

   Title_Dona  Title_Dr  Title_Jonkheer  Title_Lady  Title_Major  \
0       False     False           False       False        False   
1       False     False           False       False        False   
2       False     False           False       False        False   
3       False     False           False       False        False   
4       False     False           False       False        False   

   Title_Master  Title_Miss  Title_Mlle  Title_Mme  Title_Mr  Title_Mrs  \
0         False       False       False      False      True      False   
1         False       False       False      False     False       True   
2         False        True       False      False     False      False   
3         False       False       False      False     False       True   
4         False       False       False      False      True      False   

   Title_Ms  Title_Rev  Title_Sir  Title_the Countess  TicketGroup_A4  \
0     False      False      False               False           False   
1     False      False      False               False           False   
2     False      False      False               False           False   
3     False      False      False               False           False   
4     False      False      False               False           False   

   TicketGroup_A5  TicketGroup_AS  TicketGroup_C  TicketGroup_CA  \
0            True           False          False           False   
1           False           False          False           False   
2           False           False          False           False   
3           False           False          False           False   
4           False           False          False           False   

   TicketGroup_CASOTON  TicketGroup_FC  TicketGroup_FCC  TicketGroup_Fa  \
0                False           False            False           False   
1                False           False            False           False   
2                False           False            False           False   
3                False           False            False           False   
4                False           False            False           Fal

In [19]:
X_train.shape

(891, 83)

[go to top](#top)  

<div class="alert alert-block alert-info">
<img style="float: left" src="https://i.imgur.com/HVpiyd6.png[/img" width="60">
<b style = "font-family: Arial; font-size: 16px">Inference:</b><p style = "font-family:Verdana; font-size:14px">

<ol> We have identified features that are correlated with the survival rate on the Titanic:
    <li><b>The top 5 features positively correlated with survival are: 
        <ul>
            <li>Sex (Female) <li>Title (Mrs and Miss) <li>Fare_log <li>PClass = 1 (first class) available when string  <li>Family Group = (small)</ul>
    <li><b>The top 5 features negatively correlated with survival are:
        <ul>
            <li>Title (Mr) <li>Sex (Male)<li>Fare_log <li>PClass = 3 (third class) (no class if integer)<li>CabinGroup = unknown (missing cabin) <li>Family Size = 1<li>Family Group = (solo)</ul> 
</ol>
</div>

[go to top](#top)  

## Hyperparameter Tuning (1st)
### Random Forest

In [20]:
def model_rf_selection(X_train, y_train, cv=5, scoring='accuracy', class_weight=False):
    """
    Performs hyperparameter tuning for RandomForestClassifier using GridSearchCV.

    Parameters:
        X_train (pd.DataFrame): Training features.
        y_train (pd.Series or np.array): Training target.
        cv (int): Number of cross-validation folds (default is 5).
        scoring (str): Scoring metric (default is 'accuracy').
        class_weight (bool): Whether to apply class_weight='balanced' (default is False).

    Returns:
        GridSearchCV object (fitted)
    """
    
    # Set class_weight
    class_weight = 'balanced' if class_weight else None

    pipe_rf = Pipeline([
        ('scaler', StandardScaler()),
        ('classifier', RandomForestClassifier(random_state=42, class_weight=class_weight))
    ])

    param_grid = {
        'classifier__max_depth': [4, 6, 8, 9, 10],
        'classifier__n_estimators': [50, 100],
        'classifier__max_features': ['sqrt', 'log2'],
        'classifier__min_samples_split': [2, 3, 10],
        'classifier__min_samples_leaf': [1, 3, 10],
        'classifier__bootstrap': [True],
        'classifier__oob_score': [True, False],
        'classifier__criterion': ['gini', 'entropy', 'log_loss']
    }

    grid_search = GridSearchCV(pipe_rf, param_grid, cv=cv, scoring=scoring, n_jobs=-1)
    grid_search.fit(X_train, y_train)

    print("Best Parameters:", grid_search.best_params_)
    print(f"Best {scoring.capitalize()} Score: {grid_search.best_score_:.4f}")
    
    return grid_search

#### Revisit Model 1: Random forest, lighly imbalanced data and 83 features

In [21]:
%%time
#6min 24s
#model_rf_1 = model_rf_selection(X_train, y_train, cv=5, scoring='accuracy', class_weight=False)

CPU times: user 3 µs, sys: 1e+03 ns, total: 4 µs
Wall time: 7.63 µs


In [22]:
# Save to file after fitting
#dump(model_rf_1, 'model_rf_1.joblib')
#model_rf_1.best_estimator_

# Load it later when needed
grid_loaded = load('/kaggle/input/task001-p03-ml-models/scikitlearn/default/1/model_rf_1.joblib')

# Access best estimator
model_rf_1 = grid_loaded.best_estimator_

<h4> When feature engineering is done, we usually tend to decrease the dimensionality by selecting the "right" number of features that capture the essential.
In fact, feature selection comes with many benefits:
<ul>
    <li>It decreases redundancy among the data
    <li>It speeds up the training process
    <li>It reduces overfitting</ul>
Tree-based estimators can be used to compute feature importances, which in turn can be used to discard irrelevant features.
</h4>

In [23]:
def train_and_predict(model, X_train, y_train, X_test, test_ids, cv=5, submission_name='submission.csv'):
    """
    Trains the model, performs cross-validation, and makes predictions on the test set.

    Parameters:
    - model: Machine learning model with fit/predict methods
    - X_train: Training features
    - y_train: Training labels
    - X_test: Test features
    - test_ids: IDs for submission (e.g., PassengerId)
    - cv: Number of cross-validation folds (default=5)
    - submission_name: Filename for the CSV submission
    """
    
    # Step 1: Train the model
    model.fit(X_train, y_train)
    
    # Step 2: Cross-validation
    scores = cross_val_score(model, X_train, y_train, cv=cv)
    print("Cross-validation scores:", scores)
    print("Mean cross-validation score:", scores.mean())
    
    # Step 3: Retrain (optional here, just to match original process)
    model.fit(X_train, y_train)
    
    # Step 4: Predict
    y_pred = model.predict(X_test).astype(int)
    
    # Step 5: Create submission
    submission = pd.DataFrame({
        'PassengerId': test_ids,
        'Survived': y_pred
    })
    
    submission.to_csv(submission_name, index=False)
    print(f"Submission saved to {submission_name}")

In [24]:
def compute_metrics_df(models, X, y, cv):
    """
    Computes cross-validated metrics for a list of models and returns a sorted DataFrame.
    
    Parameters:
        models (list): List of tuples like [('Model Name', model_instance), ...]
        X (pd.DataFrame): Feature data.
        y (pd.Series or np.array): Target values.
    
    Returns:
        pd.DataFrame: DataFrame of metrics sorted by Accuracy Mean.
    """
    results = []

    for name, model in models:
        # Cross-validated accuracy scores
        accuracy_scores = cross_val_score(model, X, y, cv=cv, scoring='accuracy')

        # Cross-validated predictions and predicted probabilities
        y_pred = cross_val_predict(model, X, y, cv=cv)
        y_proba = cross_val_predict(model, X, y, cv=cv, method='predict_proba')[:, 1]

        # Collect metrics
        metrics = {
            'Model': name,
            'Accuracy Mean': np.mean(accuracy_scores),
            'Accuracy Std': np.std(accuracy_scores),
            'AUC': roc_auc_score(y, y_proba),
            'F1 Score': f1_score(y, y_pred),
            'Precision': precision_score(y, y_pred),
            'Recall': recall_score(y, y_pred)
        }
        results.append(metrics)

    # Convert to DataFrame and sort
    metrics_df = pd.DataFrame(results)
    metrics_df.sort_values(by='Accuracy Mean', ascending=False, inplace=True)
    
    return metrics_df

In [25]:
%%time
#566 ms
train_and_predict(
    model=model_rf_1, # Tweak to find the best model and try other ml models
    X_train=X_train, # Tweak the features (e.g. add feature engineering or drop feature(s), features importance )
    y_train=y_train,
    X_test=X_test, # follow the changes in X_train
    test_ids=df[891:]['PassengerId'],
    cv=5, # Tweak the cv
    submission_name='submit_rf_1.csv'
)

Cross-validation scores: [0.83798883 0.8258427  0.83707865 0.80898876 0.87640449]
Mean cross-validation score: 0.8372606867114432
Submission saved to submit_rf_1.csv
CPU times: user 832 ms, sys: 6.87 ms, total: 838 ms
Wall time: 852 ms


In [26]:
# Define your model list
models = [
    ('Random Forest', model_rf_1),
    # Add more models if needed
]

# Get the metrics DataFrame
metrics_df = compute_metrics_df(models, X_train, y_train, cv=5)
metrics_df

Model  Accuracy Mean  Accuracy Std       AUC  F1 Score  Precision  \
0  Random Forest       0.837261      0.022192  0.869862  0.778626   0.814696   

     Recall  
0  0.745614

#### Score at kaggle 78.5%

## Feature Importance
### Random Forest

In [27]:
def top_feature_importance_plot(model, X_train, top_n='pareto', classifier_step='classifier', figsize=(10, 8), save_path=None):
    """
    Plot feature importances using either Pareto principle (80% cumulative) or top N features.

    Parameters:
    - model: fitted pipeline model (e.g., Pipeline with classifier step)
    - X_train: training dataset used to retrieve feature names
    - top_n: 'pareto' to use 80% cumulative importance, or integer for top N features
    - classifier_step: name of the classifier step in pipeline (default: 'classifier')
    - figsize: size of the plot
    - save_path: if provided, saves the plot to the given file path
    """
    # Get feature names
    feature_names = X_train.columns if hasattr(X_train, 'columns') else [f'feature_{i}' for i in range(X_train.shape[1])]
    
    # Extract feature importances
    importances = model.named_steps[classifier_step].feature_importances_
    sorted_idx = np.argsort(importances)[::-1]  # descending order

    if top_n == 'pareto':
        # Compute cumulative importances
        cum_importance = np.cumsum(importances[sorted_idx])
        # Find cutoff for 80% cumulative importance
        pareto_cutoff = np.argmax(cum_importance >= 0.8) + 1
        selected_idx = sorted_idx[:pareto_cutoff]
        title_suffix = 'Top Features (Pareto 80%)'
    elif isinstance(top_n, int):
        selected_idx = sorted_idx[:top_n]
        title_suffix = f'Top {top_n} Features'
    else:
        raise ValueError("top_n must be an integer or 'pareto'")

    # Plot
    plt.figure(figsize=figsize)
    plt.barh(
        [feature_names[i] for i in selected_idx][::-1],
        importances[selected_idx][::-1],
        color='skyblue'
    )
    plt.xlabel('Gini Importance', fontsize=12)
    plt.title(f'Feature Importance - {title_suffix}', fontsize=14, pad=20)
    plt.gca().invert_yaxis()
    plt.grid(axis='x', linestyle='--', alpha=0.7)
    plt.tight_layout()
    
    # Save or show
    if save_path:
        plt.savefig(save_path, dpi=300, bbox_inches='tight')
    plt.show()

In [28]:
def reduce_features_with_importance(model, X_train, X_test, classifier_step='classifier', threshold='pareto'):
    """
    Reduces features in X_train and X_test based on feature importances using:
    - 'pareto' (default): top features that cumulatively contribute to 80% of total importance
    - 'mean': selects features above the mean importance
    - float: custom importance threshold

    Parameters:
    - model: fitted pipeline model (e.g., Pipeline with a classifier)
    - X_train: training data (DataFrame or ndarray)
    - X_test: test data (same structure as X_train)
    - classifier_step: name of the classifier step in the pipeline
    - threshold: 'pareto' (default), 'mean', float, or valid SelectFromModel threshold

    Returns:
    - X_train_reduced: training data with selected features
    - X_test_reduced: test data with selected features
    - selected_feature_names: list of retained feature names
    """
    clf = model.named_steps[classifier_step]
    importances = clf.feature_importances_

    feature_names = (
        X_train.columns if hasattr(X_train, 'columns')
        else [f'feature_{i}' for i in range(X_train.shape[1])]
    )

    if threshold == 'pareto':
        sorted_idx = np.argsort(importances)[::-1]
        sorted_importances = importances[sorted_idx]
        cumulative = np.cumsum(sorted_importances)
        pareto_cutoff = np.argmax(cumulative >= 0.8) + 1
        threshold_value = sorted_importances[pareto_cutoff - 1]
        selector = SelectFromModel(clf, threshold=threshold_value, prefit=True)
    else:
        selector = SelectFromModel(clf, threshold=threshold, prefit=True)

    # Transform the data
    X_train_reduced = selector.transform(X_train)
    X_test_reduced = selector.transform(X_test)

    # Get selected feature names
    selected_mask = selector.get_support()
    selected_feature_names = [name for name, selected in zip(feature_names, selected_mask) if selected]

    print("Reduced X_train shape:", X_train_reduced.shape)
    print("Reduced X_test shape:", X_test_reduced.shape)

    return X_train_reduced, X_test_reduced, selected_feature_names

In [29]:
X_train_reduced_1, X_test_reduced_1, selected_features = reduce_features_with_importance(model_rf_1, X_train, X_test, 
                                                                                         classifier_step='classifier', 
                                                                                         threshold='mean')
print("Selected Features:", selected_features)

Reduced X_train shape: (891, 17)
Reduced X_test shape: (418, 17)
Selected Features: ['Age', 'SibSp', 'Parch', 'CabinCnt', 'Fare_log', 'Pclass_1', 'Pclass_2', 'Pclass_3', 'Sex_female', 'Sex_male', 'CabinGroup_U', 'Title_Master', 'Title_Miss', 'Title_Mr', 'Title_Mrs', 'FamilyGroup_Large Family', 'FamilyGroup_Small Family']


X has feature names, but SelectFromModel was fitted without feature names
X has feature names, but SelectFromModel was fitted without feature names


#### Revisit Model 4: Random forest, lightly imbalanced data and 17 features

In [30]:
%%time
#5min 51s
#model_rf_3 = model_rf_selection(X_train_reduced_1, y_train, cv=5, scoring='accuracy', class_weight=False)

CPU times: user 3 µs, sys: 0 ns, total: 3 µs
Wall time: 7.63 µs


In [31]:
# Save to file after fitting
#dump(model_rf_3, 'model_rf_3.joblib')
#model_rf_3.best_estimator_

# Load it later when needed
grid_loaded = load('/kaggle/input/task001-p03-ml-models/scikitlearn/default/1/model_rf_3.joblib')

# Access best estimator
model_rf_3 = grid_loaded.best_estimator_

In [32]:
%%time
#1.5s
train_and_predict(
    model=model_rf_3, # Tweak to find the best model and try other ml models
    X_train=X_train_reduced_1, # Tweak the features (e.g. add feature engineering or drop feature(s), features importance )
    y_train=y_train,
    X_test=X_test_reduced_1, # follow the changes in X_train
    test_ids=df[891:]['PassengerId'],
    cv=5, # Tweak the cv
    submission_name='submit_rf_3.csv'
)

Cross-validation scores: [0.8547486  0.82022472 0.8258427  0.81460674 0.85955056]
Mean cross-validation score: 0.8349946644906158
Submission saved to submit_rf_3.csv
CPU times: user 1.4 s, sys: 6.7 ms, total: 1.41 s
Wall time: 1.41 s


#### Score at kaggle 79,2%

#### Continue to model 10: Random forest, lightly imbalanced data, reduced treshold to pareto

In [33]:
X_train_reduced_5, X_test_reduced_5, selected_features_5 = reduce_features_with_importance(model_rf_1, X_train, X_test, 
                                                                                         classifier_step='classifier', 
                                                                                         threshold='pareto')
print("Selected Features:", selected_features_5)

Reduced X_train shape: (891, 13)
Reduced X_test shape: (418, 13)
Selected Features: ['Age', 'CabinCnt', 'Fare_log', 'Pclass_1', 'Pclass_3', 'Sex_female', 'Sex_male', 'CabinGroup_U', 'Title_Miss', 'Title_Mr', 'Title_Mrs', 'FamilyGroup_Large Family', 'FamilyGroup_Small Family']


X has feature names, but SelectFromModel was fitted without feature names
X has feature names, but SelectFromModel was fitted without feature names


In [34]:
%%time
#5min 51s
#model_rf_10 = model_rf_selection(X_train_reduced_5, y_train, cv=5, scoring='accuracy', class_weight=False)

CPU times: user 3 µs, sys: 1e+03 ns, total: 4 µs
Wall time: 6.91 µs


In [35]:
# Save to file after fitting
#dump(model_rf_10, 'model_rf_10.joblib')
#model_rf_10.best_estimator_

# Load it later when needed
grid_loaded = load('/kaggle/input/task001-p03-ml-models/scikitlearn/default/1/model_rf_10.joblib')

# Access best estimator
model_rf_10 = grid_loaded.best_estimator_

In [36]:
%%time
#1.5s
train_and_predict(
    model=model_rf_10, # Tweak to find the best model and try other ml models
    X_train=X_train_reduced_5, # Tweak the features (e.g. add feature engineering or drop feature(s), features importance )
    y_train=y_train,
    X_test=X_test_reduced_5, # follow the changes in X_train
    test_ids=df[891:]['PassengerId'],
    cv=5, # Tweak the cv
    submission_name='submit_rf_10.csv'
)

Cross-validation scores: [0.83240223 0.80898876 0.86516854 0.79775281 0.87078652]
Mean cross-validation score: 0.835019772770071
Submission saved to submit_rf_10.csv
CPU times: user 1.42 s, sys: 2.2 ms, total: 1.42 s
Wall time: 1.42 s


#### Score at kaggle 77.3%

#### Model 11: Random forest, lightly imbalanced data, reduced treshold to 0.0075 21 Features

In [37]:
X_train_reduced_6, X_test_reduced_6, selected_features_6 = reduce_features_with_importance(model_rf_1, X_train, X_test, 
                                                                                         classifier_step='classifier', 
                                                                                         threshold=0.0075)
print("Selected Features:", selected_features_6)

Reduced X_train shape: (891, 21)
Reduced X_test shape: (418, 21)
Selected Features: ['Age', 'SibSp', 'Parch', 'CabinCnt', 'Fare_log', 'Pclass_1', 'Pclass_2', 'Pclass_3', 'Sex_female', 'Sex_male', 'Embarked_C', 'Embarked_S', 'CabinGroup_U', 'Title_Master', 'Title_Miss', 'Title_Mr', 'Title_Mrs', 'FamilyGroup_Large Family', 'FamilyGroup_Small Family', 'FamilySize_1', 'FamilySize_3']


X has feature names, but SelectFromModel was fitted without feature names
X has feature names, but SelectFromModel was fitted without feature names


In [38]:
%%time
#5min 43s
#model_rf_11 = model_rf_selection(X_train_reduced_6, y_train, cv=5, scoring='accuracy', class_weight=False)

CPU times: user 4 µs, sys: 0 ns, total: 4 µs
Wall time: 7.15 µs


In [39]:
# Save to file after fitting
#dump(model_rf_11, 'model_rf_11.joblib')
#model_rf_11.best_estimator_

# Load it later when needed
grid_loaded = load('/kaggle/input/task001-p03-ml-models/scikitlearn/default/1/model_rf_11.joblib')

# Access best estimator
model_rf_11 = grid_loaded.best_estimator_

In [40]:
%%time
#1.5s
train_and_predict(
    model=model_rf_11, # Tweak to find the best model and try other ml models
    X_train=X_train_reduced_6, # Tweak the features (e.g. add feature engineering or drop feature(s), features importance )
    y_train=y_train,
    X_test=X_test_reduced_6, # follow the changes in X_train
    test_ids=df[891:]['PassengerId'],
    cv=5, # Tweak the cv
    submission_name='submit_rf_11.csv'
)

Cross-validation scores: [0.8603352  0.82022472 0.83146067 0.80898876 0.85393258]
Mean cross-validation score: 0.8349883874207521
Submission saved to submit_rf_11.csv
CPU times: user 1.34 s, sys: 4.57 ms, total: 1.35 s
Wall time: 1.35 s


#### Score at kaggle 77.9%

#### 

#### Model 12: Random Forest with Hyperparameter tuning model_rf_3

In [41]:
# Define the pipeline with fixed parameters
model_rf_12 = Pipeline([
    ('scaler', StandardScaler()),
    ('classifier', RandomForestClassifier(
        max_depth=4,
        n_estimators=100,
        max_features='sqrt',
        min_samples_split=10,
        min_samples_leaf=1,
        bootstrap=True,
        oob_score=True,
        criterion='entropy',
        random_state=42,
        class_weight=None
    ))
])

In [42]:
# Call the function
train_and_predict(
    model=model_rf_12,
    X_train=X_train_reduced_5,
    y_train=y_train,
    X_test=X_test_reduced_5,
    test_ids=df[891:]['PassengerId'],
    cv=5,
    submission_name='submit_rf_12.csv'
)

Cross-validation scores: [0.83798883 0.82022472 0.82022472 0.80337079 0.85393258]
Mean cross-validation score: 0.8271483271608814
Submission saved to submit_rf_12.csv


#### Kaggle score 77,9%

In [43]:
# Call the function
train_and_predict(
    model=model_rf_12,
    X_train=X_train_reduced_6,
    y_train=y_train,
    X_test=X_test_reduced_6,
    test_ids=df[891:]['PassengerId'],
    cv=5,
    submission_name='submit_rf_13.csv'
)

Cross-validation scores: [0.8547486  0.82022472 0.8258427  0.80898876 0.84831461]
Mean cross-validation score: 0.8316238779737619
Submission saved to submit_rf_13.csv


#### Kaggle score 77,9%

In [44]:
# Call the function
train_and_predict(
    model=model_rf_12,
    X_train=X_train,
    y_train=y_train,
    X_test=X_test,
    test_ids=df[891:]['PassengerId'],
    cv=5,
    submission_name='submit_rf_14.csv'
)

Cross-validation scores: [0.83240223 0.82022472 0.81460674 0.7752809  0.83146067]
Mean cross-validation score: 0.8147950536689473
Submission saved to submit_rf_14.csv


#### Kaggle score 77,8%

In [45]:
# Call the function
train_and_predict(
    model=model_rf_12,
    X_train=X_train_reduced_1,
    y_train=y_train,
    X_test=X_test_reduced_1,
    test_ids=df[891:]['PassengerId'],
    cv=5,
    submission_name='submit_rf_15.csv'
)

Cross-validation scores: [0.8547486  0.82022472 0.8258427  0.81460674 0.85955056]
Mean cross-validation score: 0.8349946644906158
Submission saved to submit_rf_15.csv


We have accuracy of 85.91%, precision 81.4%

#### Kaggle score 79,2%

#### Model 13: Random Forest with Hyperparameter tuning model_rf_3

In [46]:
# Define the pipeline with fixed parameters
model_rf_13 = Pipeline([
    ('scaler', StandardScaler()),
    ('classifier', RandomForestClassifier(
        max_depth=9,
        n_estimators=200,
        max_features='sqrt',
        min_samples_split=40,
        min_samples_leaf=1,
        bootstrap=True,
        oob_score=True,
        criterion='gini',
        random_state=42,
        class_weight=None
    ))
])

In [47]:
# Call the function
train_and_predict(
    model=model_rf_13,
    X_train=X_train_reduced_1,
    y_train=y_train,
    X_test=X_test_reduced_1,
    test_ids=df[891:]['PassengerId'],
    cv=5,
    submission_name='submit_rf_16.csv'
)

Cross-validation scores: [0.84916201 0.80898876 0.82022472 0.80337079 0.85955056]
Mean cross-validation score: 0.8282593685267716
Submission saved to submit_rf_16.csv


We have accuracy of 84.91%, precision 80.33%

In [48]:
# Call the function
train_and_predict(
    model=model_rf_13,
    X_train=X_train,
    y_train=y_train,
    X_test=X_test,
    test_ids=df[891:]['PassengerId'],
    cv=5,
    submission_name='submit_rf_17.csv'
)

Cross-validation scores: [0.82122905 0.82022472 0.83146067 0.80337079 0.85393258]
Mean cross-validation score: 0.8260435628648548
Submission saved to submit_rf_17.csv


#### Get the best RF param 79.4%
#### Model 14: Random Forest with Hyperparameter tuning model_rf_3_1

In [49]:
# Load it later when needed
grid_loaded = load('/kaggle/input/best_rf_1/scikitlearn/default/11/model_rf_3_1.joblib')

# Access best estimator
model_rf_3_1 = grid_loaded.best_estimator_

In [50]:
# Define the pipeline with fixed parameters
model_rf_14 = Pipeline([
    ('scaler', StandardScaler()),
    ('classifier', RandomForestClassifier(
        max_depth=6,
        n_estimators=100,
        max_features='sqrt',
        min_samples_split=10,
        min_samples_leaf=1,
        bootstrap=True,
        oob_score=True,
        criterion='entropy',
        random_state=42,
        class_weight=None
    ))
])

In [51]:
# Call the function
train_and_predict(
    model=model_rf_14,
    X_train=X_train_reduced_1,
    y_train=y_train,
    X_test=X_test_reduced_1,
    test_ids=df[891:]['PassengerId'],
    cv=5,
    submission_name='submit_rf_18.csv'
)

Cross-validation scores: [0.83798883 0.82022472 0.8258427  0.80337079 0.85955056]
Mean cross-validation score: 0.8293955181721173
Submission saved to submit_rf_18.csv


In [52]:
X_train_reduced_1.shape

(891, 17)

In [53]:
# Call the function
train_and_predict(
    model=model_rf_14,
    X_train=X_train,
    y_train=y_train,
    X_test=X_test,
    test_ids=df[891:]['PassengerId'],
    cv=5,
    submission_name='submit_rf_19.csv'
)

Cross-validation scores: [0.83798883 0.82022472 0.8258427  0.80898876 0.85955056]
Mean cross-validation score: 0.8305191136777352
Submission saved to submit_rf_19.csv


In [54]:
X_train.shape

(891, 83)

### X_train_1 drop 'Family_size'

In [55]:
# Created dummy variablesfrom categories (also can use OneHotEncoder)
df_dummies = pd.get_dummies(df[['Pclass', 'Sex', 'Age', 'SibSp',
       'Parch', 'Embarked', 'train_set',
       'CabinGroup', 'CabinCnt', 'Fare_log', 'Title',
       'TicketGroup', 'FamilyGroup']])

X_train_1 = df_dummies[df_dummies.train_set == 1].drop(['train_set'],axis=1)
X_test_1 = df_dummies[df_dummies.train_set == 0].drop(['train_set'], axis=1)

y_train = df[df.train_set==1].Survived
y_train.shape

(891,)

In [56]:
X_train_1.shape

(891, 74)

In [57]:
# Call the function
train_and_predict(
    model=model_rf_14,
    X_train=X_train_1,
    y_train=y_train,
    X_test=X_test_1,
    test_ids=df[891:]['PassengerId'],
    cv=5,
    submission_name='submit_rf_20.csv'
)

Cross-validation scores: [0.84916201 0.82022472 0.83707865 0.80898876 0.85955056]
Mean cross-validation score: 0.8350009415604795
Submission saved to submit_rf_20.csv


We have accuracy of 84.91%, precision 80.89%

#### Model 15: Expand params

In [58]:
# Define the pipeline with fixed parameters
model_rf_15 = Pipeline([
    ('scaler', StandardScaler()),
    ('classifier', RandomForestClassifier(
        max_depth=9,
        n_estimators=40,
        max_features='sqrt',
        min_samples_split=10,
        max_leaf_nodes=None,
        min_samples_leaf=3,
        min_weight_fraction_leaf=0.0,
        n_jobs=1,
        verbose=0,
        bootstrap=True,
        oob_score=False,
        criterion='gini',
        random_state=42,
        class_weight=None,
        warm_start=False
    ))
])

In [59]:
# Call the function
train_and_predict(
    model=model_rf_15,
    X_train=X_train_1,
    y_train=y_train,
    X_test=X_test_1,
    test_ids=df[891:]['PassengerId'],
    cv=5,
    submission_name='submit_rf_21.csv'
)

Cross-validation scores: [0.82122905 0.80898876 0.84269663 0.79775281 0.85393258]
Mean cross-validation score: 0.8249199673592367
Submission saved to submit_rf_21.csv


In [60]:
# Call the function
train_and_predict(
    model=model_rf_14,
    X_train=X_train_reduced_1,
    y_train=y_train,
    X_test=X_test_reduced_1,
    test_ids=df[891:]['PassengerId'],
    cv=5,
    submission_name='submit_rf_22.csv'
)

Cross-validation scores: [0.83798883 0.82022472 0.8258427  0.80337079 0.85955056]
Mean cross-validation score: 0.8293955181721173
Submission saved to submit_rf_22.csv


In [61]:
# Call the function
train_and_predict(
    model=model_rf_15,
    X_train=X_train_1,
    y_train=y_train,
    X_test=X_test_1,
    test_ids=df[891:]['PassengerId'],
    cv=5,
    submission_name='submit_rf_23.csv'
)

Cross-validation scores: [0.82122905 0.80898876 0.84269663 0.79775281 0.85393258]
Mean cross-validation score: 0.8249199673592367
Submission saved to submit_rf_23.csv


### The best base by far (Base 83 features, reduced 17 features, lightly imbalance) - 79.2%
#### Model 12: Back to SVC

In [62]:
def model_svc_selection(X_train, y_train, cv=5, scoring='accuracy', class_weight=False):
    """
    Performs hyperparameter tuning for SVC using GridSearchCV and displays ranked results.

    Parameters:
        X_train (pd.DataFrame): Training features.
        y_train (pd.Series or np.array): Training target.
        cv (int): Number of cross-validation folds (default is 5).
        scoring (str): Scoring metric (default is 'accuracy').
        class_weight (bool): Whether to apply class_weight='balanced' (default is False).

    Returns:
        GridSearchCV object (fitted)
    """
    
    # Set class_weight
    class_weight = 'balanced' if class_weight else None

    pipe_svc = Pipeline([
        ('scaler', StandardScaler()),
        ('classifier', SVC(class_weight=class_weight, random_state=42))
    ])

    param_grid = {
        'classifier__C': [0.1, 1, 10, 100, 1000],
        'classifier__kernel': ['rbf'],
        'classifier__gamma': [0.0001, 0.001, 0.01, 0.1],
        'classifier__shrinking': [True, False],
        'classifier__tol': [1e-3, 1e-4]
    }

    grid_search = GridSearchCV(pipe_svc, param_grid, cv=cv, scoring=scoring, n_jobs=-1, return_train_score=True)
    grid_search.fit(X_train, y_train)

    print("Best Parameters:", grid_search.best_params_)
    print(f"Best {scoring.capitalize()} Score: {grid_search.best_score_:.4f}")

    # Create a DataFrame to show rank and scores
    results_df = pd.DataFrame(grid_search.cv_results_)
    
    # Use correct parameter names with 'param_' prefix
    display_cols = ['rank_test_score', 'mean_test_score', 'mean_train_score'] + \
                   [f'param_{key}' for key in param_grid.keys()]
    
    results_df = results_df.sort_values('rank_test_score')[display_cols]
    
    print("\nRanked Grid Search Results (Top 10):")
    display(results_df.head(10))
    
    return grid_search

In [63]:
%%time
#5min 51s
model_svc_1 = model_svc_selection(X_train, y_train, cv=5, scoring='accuracy', class_weight=False)

Best Parameters: {'classifier__C': 10, 'classifier__gamma': 0.0001, 'classifier__kernel': 'rbf', 'classifier__shrinking': True, 'classifier__tol': 0.001}
Best Accuracy Score: 0.8271

Ranked Grid Search Results (Top 10):


rank_test_score  mean_test_score  mean_train_score param_classifier__C  \
35                1         0.827136          0.835016                  10   
33                1         0.827136          0.835016                  10   
32                1         0.827136          0.835016                  10   
34                1         0.827136          0.835016                  10   
36                5         0.822673          0.845398                  10   
37                5         0.822673          0.845398                  10   
38                5         0.822673          0.845398                  10   
39                5         0.822673          0.845398                  10   
65                9         0.821562          0.847642                1000   
67                9         0.821562          0.847642                1000   

   param_classifier__kernel param_classifier__gamma  \
35                      rbf                  0.0001   
33                      rbf                  0.0001   
32                      rbf                  0.0001   
34                      rbf                  0.0001   
36                      rbf                   0.001   
37                      rbf                   0.001   
38                      rbf                   0.001   
39                      rbf                   0.001   
65                      rbf                  0.0001   
67                      rbf                  0.0001   

   param_classifier__shrinking param_classifier__tol  
35                       False                0.0001  
33                        True                0.0001  
32                        True                 0.001  
34                       False                 0.001  
36                        True                 0.001  
37                        True                0.0001  
38                       False                 0.001  
39                       False                0.0001  
65                        True                0.0001  
67                       False                0.0001

CPU times: user 1.08 s, sys: 211 ms, total: 1.29 s
Wall time: 23.8 s


#### Model_svc_1: 83 Features (with 10, rbf, 0.0001, False, 0.0001)

In [64]:
# Define the pipeline with fixed SVC parameters
model_svc_1 = Pipeline([
    ('scaler', StandardScaler()),
    ('classifier', SVC(
        C=10,
        kernel='rbf',
        gamma=0.0001,
        shrinking=True,
        tol=1e-4,
        random_state=42
    ))
])

In [65]:
# Call the function
train_and_predict(
    model=model_svc_1,
    X_train=X_train,
    y_train=y_train,
    X_test=X_test,
    test_ids=df[891:]['PassengerId'],
    cv=5,
    submission_name='submit_svc_1.csv'
)

Cross-validation scores: [0.84916201 0.82022472 0.82022472 0.79213483 0.85393258]
Mean cross-validation score: 0.8271357730211537
Submission saved to submit_svc_1.csv


Cross-validation scores: [0.84916201 0.82022472 0.82022472 0.79213483 0.85393258]
Mean cross-validation score: 0.8271357730211537
Submission saved to submit_svc_1.csv
CPU times: user 313 ms, sys: 997 µs, total: 314 ms
Wall time: 315 ms
Kaggle score 76.1%

#### We have accuracy of 84.91%, precision 79.21% has Kaggle score 76.1%

In [66]:
# Call the function
train_and_predict(
    model=model_svc_1,
    X_train=X_train_reduced_1,
    y_train=y_train,
    X_test=X_test_reduced_1,
    test_ids=df[891:]['PassengerId'],
    cv=5,
    submission_name='submit_svc_2.csv'
)

Cross-validation scores: [0.8547486  0.8258427  0.80898876 0.75280899 0.79775281]
Mean cross-validation score: 0.8080283723557843
Submission saved to submit_svc_2.csv


#### Kaggle score 77.5%

#### Hyperparameter tuning 1

In [67]:
# Define the pipeline with fixed SVC parameters
model_svc_2 = Pipeline([
    ('scaler', StandardScaler()),
    ('classifier', SVC(
        C=1000, # Changed from svc_1
        kernel='rbf',
        gamma=0.0001,
        shrinking=True,
        tol=1e-4,
        random_state=42
    ))
])

In [68]:
# Call the function
train_and_predict(
    model=model_svc_2,
    X_train=X_train,
    y_train=y_train,
    X_test=X_test,
    test_ids=df[891:]['PassengerId'],
    cv=5,
    submission_name='submit_svc_3.csv'
)

Cross-validation scores: [0.81005587 0.81460674 0.82022472 0.80898876 0.85393258]
Mean cross-validation score: 0.8215617349821104
Submission saved to submit_svc_3.csv


In [69]:
# Call the function
train_and_predict(
    model=model_svc_2,
    X_train=X_train_reduced_1,
    y_train=y_train,
    X_test=X_test_reduced_1,
    test_ids=df[891:]['PassengerId'],
    cv=5,
    submission_name='submit_svc_4.csv'
)

Cross-validation scores: [0.8547486  0.8258427  0.82022472 0.80898876 0.85393258]
Mean cross-validation score: 0.8327474734793799
Submission saved to submit_svc_4.csv


#### We have accuracy of 85.47%, precision 80.89%

#### Kaggle score 75.6%

### Grid search 2

In [70]:
# Define the pipeline with fixed SVC parameters
model_svc_3 = Pipeline([
    ('scaler', StandardScaler()),
    ('classifier', SVC(
        C=3, # Changed from svc_1
        kernel='rbf',
        gamma=0.0001,
        shrinking=True,
        tol=1e-4,
        random_state=42
    ))
])

In [71]:
# Call the function
train_and_predict(
    model=model_svc_3,
    X_train=X_train_reduced_1,
    y_train=y_train,
    X_test=X_test_reduced_1,
    test_ids=df[891:]['PassengerId'],
    cv=5,
    submission_name='submit_svc_5.csv'
)

Cross-validation scores: [0.81564246 0.81460674 0.80337079 0.7752809  0.81460674]
Mean cross-validation score: 0.8047015253279769
Submission saved to submit_svc_5.csv


In [72]:
# Define the pipeline with fixed SVC parameters
model_svc_4 = Pipeline([
    ('scaler', StandardScaler()),
    ('classifier', SVC(
        C=5, # Changed from svc_1
        kernel='rbf',
        gamma=0.01,
        shrinking=True,
        tol=1e-4,
        random_state=42
    ))
])

In [73]:
# Call the function
train_and_predict(
    model=model_svc_4,
    X_train=X_train_reduced_1,
    y_train=y_train,
    X_test=X_test_reduced_1,
    test_ids=df[891:]['PassengerId'],
    cv=5,
    submission_name='submit_svc_6.csv'
)

Cross-validation scores: [0.8547486  0.8258427  0.82022472 0.80337079 0.85955056]
Mean cross-validation score: 0.8327474734793799
Submission saved to submit_svc_6.csv


#### Kaggle Score 76%

In [74]:
def train_and_predict_2(model, X_train, y_train, X_test, test_ids,
                      cv=5, submission_name='submission.csv',
                      model_name='saved_model.pkl'):
    """
    Trains the model, performs cross-validation, saves the model, and makes predictions on the test set.

    Parameters:
    - model: Machine learning model with fit/predict methods
    - X_train: Training features
    - y_train: Training labels
    - X_test: Test features
    - test_ids: IDs for submission (e.g., PassengerId)
    - cv: Number of cross-validation folds (default=5)
    - submission_name: Filename for the CSV submission
    - model_name: Filename to save the trained model (default='saved_model.pkl')
    """

    # Step 1: Train the model
    model.fit(X_train, y_train)

    # Step 2: Cross-validation
    scores = cross_val_score(model, X_train, y_train, cv=cv)
    print("Cross-validation scores:", scores)
    print("Mean cross-validation score:", scores.mean())

    # Step 3: Save the trained model
    joblib.dump(model, model_name)
    print(f"Model saved to {model_name}")

    # Step 4: Predict on test set
    y_pred = model.predict(X_test).astype(int)

    # Step 5: Create submission file
    submission = pd.DataFrame({
        'PassengerId': test_ids,
        'Survived': y_pred
    })
    
    submission.to_csv(submission_name, index=False)
    print(f"Submission saved to {submission_name}")

#### Kernel = linear

In [75]:
# Define the pipeline with fixed SVC parameters
model_svc_5 = Pipeline([
    ('scaler', StandardScaler()),
    ('classifier', SVC(
        C=10,
        kernel='linear',
        #degree=2,
        gamma=0.001,
        shrinking=True,
        tol=1e-4,
        probability=True  # Enable if you want to use random_state in poly kernel
    ))
])

In [76]:
%%time
#4min 51s
# Call the function
#train_and_predict_2(
    #model=model_svc_5,
    #X_train=X_train_reduced_1,
    #y_train=y_train,
    #X_test=X_test_reduced_1,
    #test_ids=df[891:]['PassengerId'],
    #cv=5,
    #submission_name='submit_svc_7.csv',
    #model_name='model_svc_linear_1.pkl'
#)

CPU times: user 6 µs, sys: 0 ns, total: 6 µs
Wall time: 10 µs


In [77]:
# Load model from file
#loaded_model = joblib.load('/kaggle/input/best_rf_1/scikitlearn/default/12/model_svc_linear_1.pkl')

# Make predictions
#predictions = loaded_model.predict(X_test)

# If you want to display predictions directly
#print("Sample predictions:\n", predictions[:10])  # Show first 10 predictions

# If you're working with Titanic test data and have 'PassengerId':
#submission = pd.DataFrame({
    #'PassengerId': test_ids,  # Make sure you have this variable from test data
    #'Survived': predictions.astype(int)
#})

# Display head of the submission
#print("\nSubmission preview:")
#display(submission.head())

# Optionally save to CSV if needed (Kaggle output folder)
#submission.to_csv('submit_svc_linear_1.csv', index=False)

need longer time and results some similar to rbf

#### Kernel = 'poly'

Trying degree below 4 and gamma below 0.1 result below 80%

In [78]:
# Define the pipeline with fixed SVC parameters
model_svc_6 = Pipeline([
    ('scaler', StandardScaler()),
    ('classifier', SVC(
        C=5,
        kernel='poly',
        degree=4,
        gamma=0.1,
        shrinking=True,
        tol=1e-4,
        probability=True  # Enable if you want to use random_state in poly kernel
    ))
])

In [79]:
%%time
# Call the function
train_and_predict(
    model=model_svc_6,
    X_train=X_train_reduced_1,
    y_train=y_train,
    X_test=X_test_reduced_1,
    test_ids=df[891:]['PassengerId'],
    cv=5,
    submission_name='submit_svc_poly_1.csv'
)

Cross-validation scores: [0.80446927 0.78089888 0.82022472 0.78651685 0.84269663]
Mean cross-validation score: 0.8069612704789405
Submission saved to submit_svc_poly_1.csv
CPU times: user 1.12 s, sys: 0 ns, total: 1.12 s
Wall time: 1.12 s


#### Kaggle score 74.2%

#### Apply rule

In [80]:
def apply_group_rules(df):
    """
    Apply prediction rules based on demographic and family survival likelihood.
    - Predict '0' (die) for highest fatality group.
    - Predict '1' (live) for highest survival group.
    - Leave others as NaN for ML model to predict later.
    """
    # Ensure necessary columns exist
    required_cols = ['Sex', 'Title', 'AgeGroup', 'FamilyGroup']
    for col in required_cols:
        if col not in df.columns:
            raise ValueError(f"Missing required column: {col}")

    # Create a new column to store predictions
    df['Survived_Predict'] = np.nan

    # --- Rule 1: Predict Die for Highest Fatality Group ---
    cond_die = (
        (df['Sex'] == 'male') &
        (df['Title'] == 'Mr') &
        (df['AgeGroup'].isin(['Teen', 'Young Adult', 'Adult', 'Senior'])) &
        (df['FamilyGroup'].isin(['Solo', 'Small']))
    )
    df.loc[cond_die, 'Survived_Predict'] = 0

    # --- Rule 2: Predict Live for Highest Survival Group ---
    cond_live = (
        (df['Sex'] == 'female') &
        (df['Title'].isin(['Miss', 'Mrs'])) &
        (df['AgeGroup'].isin(['Adult', 'Senior'])) &
        (df['FamilyGroup'].isin(['Solo', 'Small']))
    )
    df.loc[cond_live,
    'Survived_Predict'] = 1

    return df

In [81]:
df_2 = df.copy()
# Apply rules to your dataframe
df_2 = apply_group_rules(df_2)

In [82]:
# Convert the 'Pclass' and 'FamilySize' columns to string (object) type for pd.get_dummies() Using this will acquire 83 features
# Let us apply without converting to string
df_2[['Survived_Predict']] = df_2[['Survived_Predict']].astype(str)

In [83]:
# Created dummy variablesfrom categories (also can use OneHotEncoder)
df_dummies = pd.get_dummies(df_2[['Pclass', 'Sex', 'Age', 'SibSp',
       'Parch', 'Embarked', 'train_set',
       'CabinGroup', 'CabinCnt', 'Fare_log', 'Title',
       'TicketGroup', 'FamilyGroup','FamilySize','Survived_Predict']])

X_train_2 = df_dummies[df_dummies.train_set == 1].drop(['train_set'],axis=1)
X_test_2 = df_dummies[df_dummies.train_set == 0].drop(['train_set'], axis=1)

In [84]:
X_train_2.shape

(891, 86)

In [85]:
%%time
#6min 17s
model_rf_svc = model_rf_selection(X_train_2, y_train, cv=5, scoring='accuracy', class_weight=False)

Best Parameters: {'classifier__bootstrap': True, 'classifier__criterion': 'gini', 'classifier__max_depth': 6, 'classifier__max_features': 'sqrt', 'classifier__min_samples_leaf': 1, 'classifier__min_samples_split': 2, 'classifier__n_estimators': 100, 'classifier__oob_score': True}
Best Accuracy Score: 0.8316
CPU times: user 14 s, sys: 1.09 s, total: 15.1 s
Wall time: 5min 37s


In [86]:
# Save to file after fitting
dump(model_rf_svc, 'model_rf_svc.joblib')
model_rf_svc.best_estimator_

# Load it later when needed
#grid_loaded = load('/kaggle/input/task001-p03-ml-models/scikitlearn/default/1/model_rf_10.joblib')

# Access best estimator
#model_rf_10 = grid_loaded.best_estimator_

Pipeline(steps=[('scaler', StandardScaler()),
                ('classifier',
                 RandomForestClassifier(max_depth=6, oob_score=True,
                                        random_state=42))])

In [87]:
# Use the best estimator (a pipeline) from the fitted GridSearchCV object
best_pipeline = model_rf_svc.best_estimator_

X_train_reduced_7, X_test_reduced_7, selected_features_7 = reduce_features_with_importance(
    best_pipeline,
    X_train_2,
    X_test_2,
    classifier_step='classifier',
    threshold='mean'
)

Reduced X_train shape: (891, 19)
Reduced X_test shape: (418, 19)


X has feature names, but SelectFromModel was fitted without feature names
X has feature names, but SelectFromModel was fitted without feature names


In [88]:
%%time
#6min 17s
model_svc_7 = model_svc_selection(X_train_2, y_train, cv=5, scoring='accuracy', class_weight=False)

Best Parameters: {'classifier__C': 10, 'classifier__gamma': 0.0001, 'classifier__kernel': 'rbf', 'classifier__shrinking': True, 'classifier__tol': 0.001}
Best Accuracy Score: 0.8294

Ranked Grid Search Results (Top 10):


rank_test_score  mean_test_score  mean_train_score param_classifier__C  \
35                1         0.829383          0.833613                  10   
34                1         0.829383          0.833613                  10   
33                1         0.829383          0.833613                  10   
32                1         0.829383          0.833613                  10   
55                5         0.824914          0.859428                 100   
54                5         0.824914          0.859428                 100   
52                5         0.824914          0.859428                 100   
53                5         0.824914          0.859428                 100   
21                9         0.823771          0.834455                   1   
23                9         0.823771          0.834455                   1   

   param_classifier__kernel param_classifier__gamma  \
35                      rbf                  0.0001   
34                      rbf                  0.0001   
33                      rbf                  0.0001   
32                      rbf                  0.0001   
55                      rbf                   0.001   
54                      rbf                   0.001   
52                      rbf                   0.001   
53                      rbf                   0.001   
21                      rbf                   0.001   
23                      rbf                   0.001   

   param_classifier__shrinking param_classifier__tol  
35                       False                0.0001  
34                       False                 0.001  
33                        True                0.0001  
32                        True                 0.001  
55                       False                0.0001  
54                       False                 0.001  
52                        True                 0.001  
53                        True                0.0001  
21                        True                0.0001  
23                       False                0.0001

CPU times: user 697 ms, sys: 53.8 ms, total: 751 ms
Wall time: 20.7 s


In [89]:
# Save to file after fitting
#dump(model_svc_7, 'model_svc_7.joblib')
#model_svc_7.best_estimator_

# Load it later when needed
#grid_loaded = load('/kaggle/input/task001-p03-ml-models/scikitlearn/default/1/model_rf_10.joblib')

# Access best estimator
#model_rf_10 = grid_loaded.best_estimator_

In [90]:
%%time
# Call the function
train_and_predict(
    model=model_svc_7,
    X_train=X_train_reduced_7,
    y_train=y_train,
    X_test=X_test_reduced_7,
    test_ids=df[891:]['PassengerId'],
    cv=5,
    submission_name='submit_svc_9.csv'
)

Cross-validation scores: [0.81564246 0.82022472 0.82022472 0.79213483 0.83146067]
Mean cross-validation score: 0.8159374803841567
Submission saved to submit_svc_9.csv
CPU times: user 2.91 s, sys: 214 ms, total: 3.13 s
Wall time: 1min 33s


#### Kaggle score 76.3%

#### Applied rule

[go to top](#top)  

<img style="float:left" src="https://i.imgur.com/wpcEXQC.png" width="50"><div style = "font-family: Arial; font-size: 16px">
    <h1>Execute - share</h1></div>

<h1><a id="Ex">Summary</a></h1>  

We have built classification models to predict categorical course ratings using the embedding feature vectors extracted from neural networks.
<hr>

<div class="alert alert-block alert-success" style="font-family:verdana; font-size:14px">
</ol>

In [91]:
def compare_rf_models(models_dict, kaggle_scores_dict):
    """
    Compares specified hyperparameters and Kaggle scores for a dictionary of Random Forest models.

    Parameters:
        models_dict (dict): Dictionary of model name to model object.
        kaggle_scores_dict (dict): Dictionary of model name to Kaggle score.

    Returns:
        pd.DataFrame: DataFrame comparing model parameters and Kaggle scores.
    """
    # Parameters to extract from each model
    params_to_extract = [
        'classifier__max_depth',
        'classifier__n_estimators',
        'classifier__max_features',
        'classifier__min_samples_split',
        'classifier__min_samples_leaf',
        'classifier__bootstrap',
        'classifier__oob_score',
        'classifier__criterion'
    ]
    
    model_data = []
    
    for model_name, model in models_dict.items():
        params = model.get_params()
        row = {'Model': model_name, 'Kaggle Score (%)': kaggle_scores_dict.get(model_name, None)}
        for param in params_to_extract:
            row[param] = params.get(param, None)
        model_data.append(row)


    # Create DataFrame and sort by Kaggle Score
    df = pd.DataFrame(model_data)
    df_sorted = df.sort_values(by='Kaggle Score (%)', ascending=False).reset_index(drop=True)
    
    return df_sorted

In [92]:
# Example usage:
models = {
    'model_rf_1': model_rf_1,
    'model_rf_3': model_rf_3,
     'model_rf_3_1': model_rf_3_1,
    'model_rf_10': model_rf_10,
    'model_rf_11': model_rf_11
}

kaggle_scores = {
    'model_rf_1': 78.5,
    'model_rf_3': 79.2,
    'model_rf_3_1': 79.4,
    'model_rf_10': 77.3,
    'model_rf_11': 77.9
}

# Call the function
df_rf_comparison = compare_rf_models(models, kaggle_scores)

# Display the result
df_rf_comparison

Model  Kaggle Score (%)  classifier__max_depth  \
0  model_rf_3_1              79.4                      6   
1    model_rf_3              79.2                      4   
2    model_rf_1              78.5                     10   
3   model_rf_11              77.9                      4   
4   model_rf_10              77.3                     10   

   classifier__n_estimators classifier__max_features  \
0                       100                     sqrt   
1                       100                     sqrt   
2                        50                     sqrt   
3                       100                     sqrt   
4                       100                     sqrt   

   classifier__min_samples_split  classifier__min_samples_leaf  \
0                             10                             1   
1                             10                             1   
2                              2                             3   
3                              3                             1   
4                              2                             3   

   classifier__bootstrap  classifier__oob_score classifier__criterion  
0                   True                   True               entropy  
1                   True                   True               entropy  
2                   True                   True                  gini  
3                   True                   True                  gini  
4                   True                   True                  gini

In <a href="https://www.kaggle.com/code/wahyuardhitama/task001-p01-ml-titanic-20250407">Part 2 (P02)</a>Part 1 (P01), we conducted the P and A phase. <a href="https://www.kaggle.com/code/wahyuardhitama/task001-p02-ml-titanic-20250416">Part 2 (P02)</a>, we explored 12 machine learning models, including initial tuning for Random Forest. Subsequently, part 3 (P03)</a> focused on further exploration and tuning of Random Forest and Support Vector Classifier (SVC) models and also we add some back to part 2.Finally,  <a href="https://www.kaggle.com/code/wahyuardhitama/task001-p04-ml-titanic-20250507">part 4 (P04) investigated boosting techniques and a Voting Classifier incorporating Boosting and Random Forest.</a>

In [93]:
from IPython.display import display, HTML
# Link HTML files for pages
output_file_01 = "https://www.kaggle.com/code/wahyuardhitama/task001-p01-ml-titanic-20250407"
output_file_02 = "https://www.kaggle.com/code/wahyuardhitama/task001-p02-ml-titanic-20250416"
output_file_04 = "https://www.kaggle.com/code/wahyuardhitama/task001-p04-ml-titanic-20250507"

# Display links to the saved HTML files
display(HTML(f"<p style='font-size: 18px;'>Go to <a href='{output_file_01}' target='_blank' style='font-size: 18px;'>P01 </a></p>"))
display(HTML(f"<p style='font-size: 18px;'>Go to <a href='{output_file_02}' target='_blank' style='font-size: 18px;'>P02 </a></p>"))
display(HTML(f"<p style='font-size: 18px;'>Go to <a href='{output_file_04}' target='_blank' style='font-size: 18px;'>P04 </a></p>"))

[go to top](#top)  